### PR TITLE
feat: add go protobuf generation

### PR DESCRIPTION
**Pull request by zyy17**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add go protobuf generation.

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x]  I have written the necessary comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by zyy17** [comment]
Oops... I forget to add  `proto-gen-go` in protoc installation, and I will add it ASAP.

File: scripts/generate-go.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+          --go-grpc_out=${GO_OUTPUT}  --go-grpc_opt=paths=source_relative \
+          -Iproto proto/greptime/v1/meta/*.proto
+
+${PROTOC} --go_out=${GO_OUTPUT}  --go_opt=paths=source_relative \
+          --go-grpc_out=${GO_OUTPUT}  --go-grpc_opt=paths=source_relative \
+          -Iproto proto/greptime/v1/*.proto
+
+${PROTOC} --go_out=${GO_OUTPUT}  --go_opt=paths=source_relative \
+          --go-grpc_out=${GO_OUTPUT}  --go-grpc_opt=paths=source_relative \
+          -Iproto proto/prometheus/remote/*.proto
```
Comment (yuanbohan) [inline scripts/generate-go.sh:24]: Is it helpful to generate the code via temporary docker container? like https://github.com/namely/docker-protoc

File: scripts/install-protoc.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      echo "Downloading Linux protoc ${PROTOC_VERSION} for x86_64"
+      curl  ${DOWNLOAD_URL}/protoc-${PROTOC_VERSION}-linux-x86_64.zip -o ${LOCAL_DIR}/protoc.zip -L
+    fi
+else
+    echo "Unsupported platform"
+    exit 1
+fi
+
+echo "Unzipping protoc..."
+unzip ${LOCAL_DIR}/protoc.zip -d ${LOCAL_DIR}/protoc
```
Comment (yuanbohan) [inline scripts/install-protoc.sh:43]: Is it needed to add the dependencies listed in this script file into README?

File: README.md [hunk 3]
```diff
@@ -34,6 +52,23 @@ use greptime_proto::v1::meta::*;
...
+
+```console
+go get github.com/GreptimeTeam/greptime-proto@main
+```
+
+Then use greptime-proto as the normal Go module:
+
+```go
+import (
+        greptimev1 "github.com/GreptimeTeam/greptime-proto/go/greptime/v1"
```
Comment (yuanbohan) [inline README.md:67]: Is the package path verbose? like there are three `greptime` in the package path

File: scripts/generate-go.sh (hunk 1 above)
Comment (zyy17) [reply scripts/generate-go.sh:24]: It's excellent advice! I will change to your advice later.

Comment (vinland-avalon) [reply scripts/generate-go.sh:24]: Or maybe we can make this progress more automatical? It helps if the updates of proto files is just followed by the updates of generated files in this repo. Github Actions may help.

File: README.md (hunk 3 above)
Comment (zyy17) [reply README.md:67]: I think it's the style issue. For now, the package path is consistent with the directory structure of `./proto`.

For example:

The `meta/*.proto` is in `./proto/greptime/v1/meta/`, then the generated `meta` go package will use `${ModuleRoot}/go/greptime/v1` as its path.

File: scripts/generate-go.sh (hunk 1 above)
Comment (zyy17) [reply scripts/generate-go.sh:24]: yeap, it's also very helpful, but I think we can make it in the next PR(GitHub Action is very very hard to debug...).

File: Makefile [hunk 4]
```diff
@@ -1,6 +1,21 @@
...
+	go mod download
+
+install-protoc:
+	./scripts/install-protoc.sh
+
+install-protobuf-go-plugins:
+	./scripts/install-protobuf-go-plugins.sh
+
+clean:
+	rm -rf bin
```
Comment (MichaelScofield) [inline Makefile:21]: cautious for `rm -rf / bin`

```suggestion
	rm -r bin
```

File: Makefile [hunk 5]
```diff
@@ -1,6 +1,21 @@
...
 rust:
 	cargo build
+
+go: install-protoc install-protobuf-go-plugins go-deps
+	./scripts/generate-go.sh
+
+go-deps:
+	go mod download
+
+install-protoc:
```
Comment (MichaelScofield) [inline Makefile:14]: I suggest not to install protoc for users, at least not include it in our makefile. Users should be able to use their prefer way to install protoc. If we place protoc installation here, we are very likely to carry the burden to maintain it, which is tedious and unnecessary.

File: scripts/generate-go.sh (hunk 1 above)
Comment (zyy17) [reply scripts/generate-go.sh:24]: @yuanbohan When I try `namely/docker-protoc`, I think it's more complicated than what I thought. To make everything more simple and controllable, I think the fixed-version binary will be better.

Comment (zyy17) [reply scripts/generate-go.sh:24]: @yuanbohan, my mistake, I figure out how to use `namely/docker-protoc`. I decide to use it.

File: Makefile (hunk 4 above)
Comment (zyy17) [reply Makefile]: I use the container `namely/protoc-all:1.51_1` to compile the protobuf codes and remove the installation.

File: Makefile (hunk 5 above)
Comment (zyy17) [reply Makefile]: You are right. As @yuanbohan suggests, I use the container `namely/protoc-all:1.51_1` to compile the code.

**Review by MichaelScofield — APPROVED** [review]

File: proto/prometheus/remote/remote.proto [hunk 6]
```diff
@@ -14,7 +14,7 @@
 syntax = "proto3";
 package prometheus;
 
-option go_package = "prompb";
+option go_package = "github.com/GreptimeTeam/greptime-proto/go/prometheus/remote";
```
Comment (yuanbohan) [inline proto/prometheus/remote/remote.proto:17]: Is it needed to update the License comment in `prometheus/remote/*.proto`, appending Greptime Team? like: 

```
Copyright 2016 Prometheus Team
Copyright 2023 Greptime Team
```

Comment (zyy17) [reply proto/prometheus/remote/remote.proto:19]: I add the comments to describe our modification.

**Review by yuanbohan — APPROVED** [review]

**Pull request merged**